keps/sig-auth/1688-dynamic-authentication-config/README.md

```diff
@@ -0,0 +1,620 @@
+`metadata.name` of the `AuthenticationConfig` object that asserted the `user.Info`.
+
+The `authenticationconfigs.authentication.k8s.io/type` key will be set to the
+`spec.type` of the `AuthenticationConfig` object that asserted the `user.Info`.
+
+More audit annotations may be added in the future to aid in tracing the source
+of the asserted identity.  For example, with the `oidc` type it may be helpful
+to include the issuer URL and client ID.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable.  This may include API specs (though not always
+required) or even code snippets.  If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+The proposed functionality will be gated behind a new feature flag called
+`DynamicAuthenticationConfig`.
+
+The proposed API will reside in the `authentication.k8s.io` group at version
+`v1alpha1` and resource `authenticationconfigs`.  The `AuthenticationConfig`
+`Kind` is defined as:
+
+```golang
+type AuthenticationConfig struct {
+  metav1.TypeMeta `json:",inline"`
+  // +optional
+  metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+  Spec AuthenticationConfigSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+}
+
+type AuthenticationConfigSpec struct {
+  Type AuthenticationConfigType `json:"type" protobuf:"bytes,1,opt,name=type"`
+
+  // this generic username and group prefix config is always required
+  // but the fields can be explicitly set to "-" to override it
+  Prefix *PrefixConfig `json:"prefix" protobuf:"bytes,2,opt,name=prefix"`
+
+  X509 *X509Config `json:"x509,omitempty" protobuf:"bytes,3,opt,name=x509"`
+
+  OIDC *OIDCConfig `json:"oidc,omitempty" protobuf:"bytes,4,opt,name=oidc"`
+
+  // This struct will be defined in a later revision
+  // Webhook *WebhookConfig `json:"webhook,omitempty" protobuf:"bytes,5,opt,name=webhook"`
+}
+
+type AuthenticationConfigType string
+
+const (
+  AuthenticationConfigTypeX509    AuthenticationConfigType = "x509"
+  AuthenticationConfigTypeOIDC    AuthenticationConfigType = "oidc"
+  AuthenticationConfigTypeWebhook AuthenticationConfigType = "webhook"
+)
+
+type X509Config struct {
+  // caBundle is a PEM encoded CA bundle used for client auth (x509.ExtKeyUsageClientAuth).
+  // +listType=atomic
+  // Required
+  CABundle []byte `json:"caBundle" protobuf:"bytes,1,opt,name=caBundle"`
+}
+
+type OIDCConfig struct {
+  // this type is defined in admission registration
+  Issuer admissionregistrationv1.WebhookClientConfig `json:"issuer" protobuf:"bytes,1,opt,name=issuer"`
+
+  ClientID string `json:"clientID" protobuf:"bytes,2,opt,name=clientID"`
+
+  UsernameClaim string `json:"usernameClaim" protobuf:"bytes,3,opt,name=usernameClaim"`
+
+  GroupsClaim string `json:"groupsClaim" protobuf:"bytes,4,opt,name=groupsClaim"`
+}
+
+type PrefixConfig struct {
```

Comment:
Rather than a prefix, maybe use a glob pattern (please not a regex)? E.g. if I want to authenticate `*@example.com` or `*@*.example.com`

Reply:
This sounds like a different use case than `PrefixConfig` which is meant to disambiguate the same username/group from different authenticators.

I think the use case is valid because it makes it significantly easier to use something like `gitlab.com` as your `oidc` IDP (i.e. maybe you do not want every user with a GitLab account to be able to authenticate to your cluster even if they are not authorized to do anything).

The approach I have been thinking around such a use case is:

```go
// the asserted user must be a member of at least one of these groups to authenticate.
// set to ["*"] to disable this check.
// required (cannot be empty)
requiredGroups []string
```

Basically, a simple gate on group membership for the `x509` and `oidc` authenticators.  `webhook` can internally do anything to limit authentication based on other parameters (i.e. some form of authz check) so it matters a lot less in that case.